File: 15. String compare via pointers.cpp
// Topic: Compare two strings using pointers (return 0 if equal)
#include <iostream>
using namespace std;

int strcmp_ptr(const char* a, const char* b)
{
    while (*a && (*a == *b))
    {
        ++a;
        ++b;
    }
    return (unsigned char)*a - (unsigned char)*b;
}

int main()
{
    const char* s1 = "abc";
    const char* s2 = "abc";
    cout << (strcmp_ptr(s1, s2) == 0 ? "Equal" : "Not equal") << '\n';
    return 0;
}

// Time Complexity: O(n) where n is length of shorter string.
// Space Complexity: O(1).

File: 15. String compare via pointers.py
"""
Topic: Compare two strings using pointer-like iteration
Description: Return 0 if equal
"""
def strcmp_ptr(a, b):
    if len(a) != len(b):
        return 1
    for x, y in zip(a, b):
        if x != y:
            return ord(x) - ord(y)
    return 0

def main():
    print('Equal' if strcmp_ptr('abc','abc')==0 else 'Not equal')

if __name__ == '__main__':
    main()

# Time Complexity: O(n).
# Space Complexity: O(1).
